### PR TITLE
feat(connector): add ingest timestamps

### PR DESCRIPTION
**Pull request by hekike**
## Why

Make it easier to debug ingested events by adding timestamps about the time of the event received and stored.

## What

Add `ingested_at` and `stored_at` timestamps to `om_events` table and `IngestedEvent` API model.

## Breaking Changes

Changes to the ClickHouse schema. To migrate execute the following manually:

```
ALTER TABLE openmeter.om_events ADD COLUMN ingested_at DateTime AFTER data;
ALTER TABLE openmeter.om_events ADD COLUMN stored_at DateTime AFTER ingested_at;
```

### CODE REVIEW

File: api/openapi.yaml [hunk 1]
```diff
@@ -1958,29 +1958,43 @@ components:
...
         validationError:
           type: string
-          readOnly: true
           example: "invalid event"
+        ingestedAt:
+          type: string
+          format: date-time
+          description: The date and time the event was ingested.
+          example: "2024-01-01T00:00:00Z"
+        createdAt:
```
Comment (sagikazarmark) [inline api/openapi.yaml:1983]: Maybe call this stored_at (based on the description)

File: openmeter/streaming/clickhouse_connector/query_test.go [hunk 2]
```diff
@@ -18,7 +18,7 @@ func TestCreateEventsTable(t *testing.T) {
 			data: createEventsTable{
 				Database: "openmeter",
 			},
-			want: "CREATE TABLE IF NOT EXISTS openmeter.om_events (namespace String, validation_error String, id String, type LowCardinality(String), subject String, source String, time DateTime, data String) ENGINE = MergeTree PARTITION BY toYYYYMM(time) ORDER BY (namespace, time, type, subject)",
+			want: "CREATE TABLE IF NOT EXISTS openmeter.om_events (namespace String, validation_error String, id String, type LowCardinality(String), subject String, source String, time DateTime, data String, ingested_at DateTime, create_at DateTime) ENGINE = MergeTree PARTITION BY toYYYYMM(time) ORDER BY (namespace, time, type, subject)",
```
Comment (sagikazarmark) [inline openmeter/streaming/clickhouse_connector/query_test.go:21]: typo: create_at?

File: openmeter/sink/storage.go [hunk 3]
```diff
@@ -58,14 +59,30 @@ func (q InsertEventsQuery) ToSQL() (string, []interface{}, error) {
...
+	query.Cols("namespace", "validation_error", "id", "type", "source", "subject", "time", "data", "ingested_at", "created_at")
 
 	for _, message := range q.Messages {
 		var eventErr string
 		if message.Status.Error != nil {
 			eventErr = message.Status.Error.Error()
 		}
 
+		ingestedAt := time.Now()
+		createdAt := time.Now()
```
Comment (sagikazarmark) [inline openmeter/sink/storage.go:71]:
```suggestion
		createdAt := time.Now()
		ingestedAt := createdAt
```

Time is immutable, it's safe to copy it and it makes sure the monotonic clock remains the same.

File: openmeter/sink/storage.go [hunk 4]
```diff
@@ -58,14 +59,30 @@ func (q InsertEventsQuery) ToSQL() (string, []interface{}, error) {
...
+
+		for _, header := range message.KafkaMessage.Headers {
+			// Parse ingested_at header
+			if header.Key == "ingested_at" {
+				var err error
+
+				ingestedAtStr := string(header.Value)
+				ingestedAt, err = time.Parse(time.RFC3339, ingestedAtStr)
```
Comment (sagikazarmark) [inline openmeter/sink/storage.go:79]:
```suggestion
				ingestedAt, err = time.Parse(time.RFC3339, string(header.Value))
```

Comment (sagikazarmark) [reply openmeter/sink/storage.go:79]: Note: this essentially resets `ingestedAt` to the zero value returned by `time.Parse`.

File: api/openapi.yaml (hunk 1 above)
Comment (hekike) [reply api/openapi.yaml:1983]: I was thinking about it, but `created_at` feels like the standard for DB modelling.

Comment (sagikazarmark) [reply api/openapi.yaml:1983]: For a standard CRUD operation, sure. This isn't that. Even the description says it's the timestamp for storing the event. Lastly, it appears in the API as well. I would prioritize ubiquitous language that makes sense in the context over DB modeling lingo.

Comment (hekike) [reply api/openapi.yaml]: Renamed it.

**Review by chrisgacsal — APPROVED** [review]

**Review by sagikazarmark — APPROVED** [review]

**Review by tothandras — APPROVED** [review]

**Pull request merged**